Entity Template Documentation Updates (#682)
spatialos/gdk-for-unity: docs/projects/fps/tutorial.md

{
    var gameLogic = WorkerUtils.UnityGameLogic;

    var healthPickupComponent = new Pickups.HealthPickup.Snapshot { IsActive = true, HealthValue = healthValue };

    var entityTemplate = new EntityTemplate();

>     entityTemplate.AddComponent(new Position.Snapshot { Coords = new Coordinates(position.X, position.Y, position.Z) }, gameLogic);

Reviewer: `position.ToSpatialCoordinates()` ?

Author: See above :P

    entityTemplate.AddComponent(new Metadata.Snapshot { EntityType = "HealthPickup"}, gameLogic);
    entityTemplate.AddComponent(new Persistence.Snapshot(), gameLogic);
    entityTemplate.AddComponent(healthPickupComponent, gameLogic);

>     entityTemplate.SetReadAccess(gameLogic, WorkerUtils.UnityClient);

Reviewer: needs write access as well?

Author: What needs write access?

Author: Discussed offline - EntityAcl

    return entityTemplate;
}
```

> You may notice that `Position` and `Metadata` appear in the entity template of _every_ entity type. This is because these are mandatory "well-known components" that SpatialOS expects. `Persistence` is another "well-known component" that is optional. 

Reviewer: do we have a page that we can link to to explain well-known components?

Author: Hmmm maybe on the main docs site

Author: Looks like we don't have anything to link to :/

{
    var gameLogic = WorkerUtils.UnityGameLogic;

    var healthPickupComponent = new Pickups.HealthPickup.Snapshot { IsActive = true, HealthValue = healthValue };

    var entityTemplate = new EntityTemplate();
    entityTemplate.AddComponent(new Position.Snapshot { Coords = new Coordinates(position.X, position.Y, position.Z) }, gameLogic);
    entityTemplate.AddComponent(new Metadata.Snapshot { EntityType = "HealthPickup"}, gameLogic);
    entityTemplate.AddComponent(new Persistence.Snapshot(), gameLogic);
    entityTemplate.AddComponent(healthPickupComponent, gameLogic);
    entityTemplate.SetReadAccess(gameLogic, WorkerUtils.UnityClient);
    entityTemplate.SetComponentWriteAccess(EntityAcl.ComponentId, gameLogic);

    return entityTemplate;
}
```

> You may notice that `Position` and `Metadata` appear in the entity template of _every_ entity type. This is because these are mandatory "well-known components" that SpatialOS expects. `Persistence` is another "well-known component" that is optional.

Reviewer 2: Have we just made up a term "well-known component"?  Do we need to add this to our terms?

Reviewer 2: (Appreciate it was already here!)

> You may notice that `Position` and `Metadata` appear in the entity template of _every_ entity type. This is because these are mandatory "well-known components" that SpatialOS expects. `Persistence` is another "well-known component" that is optional.

Reviewer 2: ```suggestion
You may notice that `Position` and `Metadata` appear in the entity template of _every_ entity type. This is because these are standard-library components that SpatialOS expects. (Note that `Persistence` is another standard-library component but this is optional.)
```
Suggest removing mandatory as it is easily confused with `required components`. You've used "SpatialOS expects" so it's clear we need it.

Author: Done!

#### Add components

From your `HealthComponent`, the GDK has generated a `Pickups.HealthPickup.Snapshot` struct.

An instance of this struct can then be added to the `HealthPickup` entity using the line: `.AddComponent(healthPickupComponent, gameLogic)`.

#### Set permissions (ACLs)

[Access Control Lists]({{urlRoot}}/content/glossary#access-control-list-acl) are how SpatialOS specifies which workers have permission to read-from or write-to the values of certain components. There may be data which you want to be kept private to server-side workers (because clients might use that knowledge to cheat!). Some components should definitely restrict their write-access to specific workers (e.g. a particular player's client) or to server-side workers only, to prevent exploits. For example, in an RPG a player should probably not be able to update the amount of gold they are carrying (at least, not without the server-side validating they aren't cheating!).

> In the EntityTemplate syntax, the `.SetReadAccess(gameLogic, WorkerUtils.UnityClient)` function call stated that all worker types should be able to read the data for this entity.

Reviewer: also need an explanation for writeaccess?

Author: Added 👍

#### Add components

From your `HealthComponent`, the GDK has generated a `Pickups.HealthPickup.Snapshot` struct.

> An instance of this struct can then be added to the `HealthPickup` entity using the line: `.AddComponent(healthPickupComponent, gameLogic)`.

Reviewer 2: ```suggestion
To add an instance of this struct to the `HealthPickup` entity, use the line: `.AddComponent(healthPickupComponent, gameLogic)`.
```

Made it active voice (I know it was passive voice there already).

[Access Control Lists]({{urlRoot}}/content/glossary#access-control-list-acl) are how SpatialOS specifies which workers have permission to read-from or write-to the values of certain components. There may be data which you want to be kept private to server-side workers (because clients might use that knowledge to cheat!). Some components should definitely restrict their write-access to specific workers (e.g. a particular player's client) or to server-side workers only, to prevent exploits. For example, in an RPG a player should probably not be able to update the amount of gold they are carrying (at least, not without the server-side validating they aren't cheating!).

In the EntityTemplate syntax, the `.SetReadAccess(gameLogic, WorkerUtils.UnityClient)` function call stated that all worker types should be able to read the data for this entity.

For each of the other components, such as your newly added `HealthPickup` component, the worker type which is given write-access is specified as a second argument to the component-adding function, e.g. `WorkerUtils.UnityGameLogic`. This is simply a string which identifies which worker type should be granted the relevant permission.

> The `EntityTemplate` also provides methods for manipulating the `ACL` of an entity without adding a snapshot. For example, `.SetComponentWriteAccess(EntityAcl.Id, gameLogic)` will set the write access for the `EntityAcl` component.

Reviewer 2: ```suggestion
The `EntityTemplate` also provides methods for manipulating the `ACL` of an entity without adding a snapshot. For example, `.SetComponentWriteAccess(EntityAcl.Id, gameLogic)` sets the write access for the `EntityAcl` component.
```
Don't need predictive future "will" - this always happens so we can use present simple.

Author: Done!

Once an entity template function exists you have a way of constructing the _template_ of an entity. You now have a couple of ways of adding a health pack entity to the world:

> * At runtime, by passing an `EntityTemplate` object to an entity creation function.

Reviewer 2: ```suggestion
* At Runtime, by passing an `EntityTemplate` object to an entity creation function.
```

Author: This doesn't refer to the `SpatialOS Runtime` so should this be capitalized? This is just the general program runtime (or execution time)

Author: 💥
